### DISCUSSION TITLE
Native Image Committer Community Meeting 2021-11-18

### DISCUSSION THREAD

**Discussion by christianwimmer** [oracle/graal#4031] [Show and tell]
List of all past and upcoming meetings: #3933

## New and Noteworthy

* Feature freeze for GraalVM 22.0 is on November 26

* JDK 8 support is going to be deleted. Working through the whole GraalVM gate system to remove JDK 8 gates.
  e02fcc15655

* JDK 17 improvements / fixes
  091e9c1a73f
  2a20d63cd1b
  99c9d7e1251
  c05d660bb64
  
* Module system support:
  #3748
  #3821
  
* Robustness:
  f191598692d
  
* Compatibility improvements:
  #3883
  #4001
  43977159498
  #3680
  #3956
  #3557
  2ad587c401f

* JFR support:
  99b234fcfdb6acf0659315360f6c86c463630f37

* Image size reduction:
  213b71cd2ef
  95305813325
  7b5f83119a8a216a72924aeaa6c116a877cd020e
  
* Cleanups:
  #3271
  
* Medium-term planning:
  New user-friendly build output: #3955
  Continuations/coroutines (support for Project Loom). Initial support was merged a year ago, but several bigger pieces and optimizations are still missing.

## Deep Dive: Native Image Serial GC policy and performance

Native Image memory management manual including tunables: https://www.graalvm.org/reference-manual/native-image/MemoryManagement/

### Serial GC Overview

* Single-threaded generational copying GC
* The object heap consists of chunks (including the image heap)
  * Same-sized chunks store small to medium objects (default: 1 MB chunks)
    * allocated as needed and cached for reuse
  * Variably-sized chunks store a single large object (currently only arrays)
* A collection copies the reachable small to medium-sized objects to new chunks
  * For large objects, their chunk is simply reassigned to the target space
  * Image heap objects are never copied or reassigned
  * In the worst case, GC needs twice the heap size while copying
* Each chunk has a card table (remembered set)
  * Entries indicate whether a memory range contains a reference
    * from the old generation to the young generation
    * from the image heap to the runtime heap
  * Card table entries are updated when writing such a reference
  * Enables collections to scan only objects in those memory ranges
* Collection policy decides
  * When to collect https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/CollectionPolicy.java#L116
  * Whether to do an young collection or a complete collection https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/CollectionPolicy.java#L126
  * The (changing) sizes of the different spaces of the heap and other parameters https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/CollectionPolicy.java#L146-L147

### Old GC policy: BySpaceAndTime

https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/BasicCollectionPolicies.java#L232

* The maximum heap size is 80% of the physical memory size
* The young generation size is 10% of the maximum heap size, but at most 256 MB
* The young generation consists only of newly created objects ("eden", "nursery") which are promoted to the old generation
  * There are no survivor spaces to delay promotion
* Collections are triggered when the young generation is full
* _Complete_ collections are triggered
  https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/BasicCollectionPolicies.java#L239-L240
  * If promoting the entire young generation would exceed the maximum heap size
  * or if more than 50% of GC time has been spent on incremental collections

### New GC policy: Adaptive

https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AdaptiveCollectionPolicy.java#L45

* The maximum heap size is still 80% of the physical memory size
* The young generation size is at most 33% of the maximum heap size
  * Eden space for newly created objects
  * Survivor spaces for young collection survivors (at most 33% of young generation)
* Collections are triggered when _eden_ is full
* _Complete_ collections are triggered
  https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AdaptiveCollectionPolicy.java#L151
  * only following a young collection: reduces peak memory usage because unreachable object are already freed
  * If the old generation is full after the young collection
  * or if the old generation is likely to become full after the next young collection
  * or if the time spent on young collections since the last complete collection is more than the time two complete collections would take (ensure old garbage does not linger for too long)
* Space sizes are continuously adjusted for throughput and footprint
  * Throughput: percentage of wall-clock time the application spends _not in_ garbage collection (cost: percentage spent _in_ GC)
  * Based on HotSpot's ParallelGC policy, but since our collector is single-threaded, we cannot expect to reach a fixed throughput goal such as 98% for multi-threaded applications
  * Spaces start at a fixed capacity: 64 MB each for the young and old generation
  * The policy keeps track of the cost of young collections and complete collections (and other statistics)
    https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AdaptiveCollectionPolicy.java#L399
  * Costs for a young/complete collection at a specific eden/old generation size are modeled using least-squares regression with exponential discounting of older data points
    https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/ReciprocalLeastSquareFit.java#L36
  * Eden space and old generation are expanded unless the models predict cost reductions below a certain threshold for doing so
    https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AdaptiveCollectionPolicy.java#L240
    https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AdaptiveCollectionPolicy.java#L452
  * Survivor space size and tenuring threshold (number of young collections after which an object is promoted to the old generation) are adjusted to balance young and complete collection costs
    https://github.com/oracle/graal/blob/9d6d87db24bfe8edf7c72e46d236d2a83a67928a/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AdaptiveCollectionPolicy.java#L203

### Future Improvements

Main issue: https://github.com/oracle/graal/issues/2386

* Variably-sized TLABs: currently fixed-size chunks are used as TLABs. Many threads but low allocation rates result in significant fragmentation and premature collections
* Mark&Compact GC for the old generation to avoid memory overhead during copying
* Optimized object pinning
* Exact write barriers especially for large arrays
* Clustering of related image heap objects
* ...

## Possible deep dive topics for next meeting
Please send suggestions, or "upvote" a suggestion, by adding a comment to this discussion.
* Reflection configuration: "query only" vs. "allow invocation" configuration for methods
* Conditional reflection configuration
* Static analysis: planned changes to image heap snapshotting
* Module system implementation: state and implications

**Comment by zakkak** [comment]
Hi, I would be interested to learn more about the planned changes to image heap snapshoting in the static analysis. Thank you Christian!

**Reply by christianwimmer** [reply]
Sounds good, talking about heap snapshotting is confirmed for the next meeting.
